torch/distributed/elastic/supervisor/launchers.py

```diff
@@ -0,0 +1,339 @@
+        conf: configuration
+    """
+    logger.info("Running torchrun supervisor policy, conf: %s", conf)
+
+    assert 0 < conf.job.min_nodes <= conf.job.max_nodes
+    assert conf.job.max_restarts >= 0
+    if conf.proc.run_path:
+        raise NotImplementedError(
+            "supervisor based torchun does not support run_path option"
+        )
+
+    connected_hosts = _allocate_hosts(ctx, conf)
+    if len(connected_hosts) < conf.job.min_nodes:
+        raise TimeoutError(
+            "Only %s out of required %s hosts were allocated",
+            len(connected_hosts),
+            conf.job.min_nodes,
+        )
+
+    failure: Optional[ProcessFailure] = None
+    for attempt in range(conf.job.max_restarts + 1):
+        logger.info(
+            "Starting attempt idx %s on hosts: %s",
+            attempt,
+            ",".join([h.hostname().result() for h in connected_hosts]),
+        )
+
+        scripts_path = str(Path(__file__).parent.absolute())
+
+        proc = [
+            sys.executable,
+            f"{scripts_path}/scripts/supervisor_prefix.py",
+            "-p",
+            "[trainer{LOCAL_RANK}]: ",
+        ]
+        proc = []
+        if not conf.proc.no_python:
+            proc = [os.getenv("PYTHON_EXEC", sys.executable), "-u"]
+            if conf.proc.module:
+                proc.append("-m")
+
+        proc += [
+            conf.job.training_script,
+            *conf.job.training_script_args,
+        ]
+
+        pg: Sequence[Process] = ctx.create_process_group(
+            list(connected_hosts),
+            args=proc,
+            processes_per_host=int(conf.job.nproc_per_node),
+            env={
+                **os.environ,
+                "TORCHELASTIC_RESTART_COUNT": str(attempt),
+                "TORCHELASTIC_MAX_RESTARTS": str(conf.job.max_restarts),
+                "MASTER_ADDR": socket.getfqdn(),
+                "MASTER_PORT": str(get_free_port()),
+                "TORCHELASTIC_RUN_ID": f"{conf.run_id}",
+                "TORCHELASTIC_ERROR_FILE": "/var/logs/templog",  # TODO
+            },
+            name=f"train/{str(attempt).zfill(3)}/{{rank}}",
+        )
+
+        failure = None
+        proc_table: Dict[Future, Process] = {p.returncode(): p for p in pg}
+        for returncode_fut in as_completed(proc_table.keys()):  # type: ignore[arg-type]
+            try:
+                r = returncode_fut.result()
+                if r != 0:
+                    failure = ProcessFailure(proc_table[returncode_fut])
+                    break
+            except Exception as e:
+                failure = ProcessFailure(proc_table[returncode_fut])
+                break
+
+        if failure:
+            for p in pg:
+                if not p.returncode().done():
+                    p.signal()
+
+            # wait for pg to clean-up within a grace period
+            logger.warning("Stopping other processes due to failure %s", attempt)
+            as_completed(
```

Comment:
as_completed is a generator so it isn't going to do anything if it isn't looped over. But maybe more generally, the supervisor `shutdown` path isn't going to exit until it's delivered its messages, so there is little benefit in waiting here.

Reply:
Thanks @zdevito. There are some CI failures + your feedback, once resolved I can merge this then.